File: windows/WeChatExporter.Windows/Services/AppSettings.cs
using System.IO;
using System.Text.Json;
using WeChatExporter.Models;

namespace WeChatExporter.Services;

public enum AppearanceMode
{
    System,
    Light,
    Dark
}

public static class AppearanceModeInfo
{
    public static string Title(AppearanceMode mode) => mode switch
    {
        AppearanceMode.Light => "浅色",
        AppearanceMode.Dark => "深色",
        _ => "跟随系统"
    };
}

public enum DateRangePreset
{
    All,
    Last7Days,
    Last30Days,
    Last90Days,
    Last365Days,
    Custom
}

public static class DateRangePresetInfo
{
    public static string Title(DateRangePreset preset) => preset switch
    {
        DateRangePreset.Last7Days => "最近 7 天",
        DateRangePreset.Last30Days => "最近 30 天",
        DateRangePreset.Last90Days => "最近 90 天",
        DateRangePreset.Last365Days => "最近 1 年",
        DateRangePreset.Custom => "自定义",
        _ => "全部时间"
    };
}

/// <summary>应用设置（持久化到 %AppData%/WeChatExporter/settings.json）。</summary>
public sealed class AppSettings
{
    public const string CreditLine = "@林琝淏科技集团有限公司出品";

    private static readonly Lazy<AppSettings> SharedLazy = new(Load);
    public static AppSettings Shared => SharedLazy.Value;

    public AppearanceMode Appearance { get; set; } = AppearanceMode.System;
    public ExportStyle ExportStyle { get; set; } = ExportStyle.SingleHtml;
    public bool IncludeMedia { get; set; }
    public bool IncludeStickerGallery { get; set; } = true;
    public bool FolderIncludeCsv { get; set; } = true;
    public bool FolderIncludeJson { get; set; }
    public bool OpenFolderAfterExport { get; set; }
    public string ExportPath { get; set; } = DefaultExportPath();
    public DateRangePreset DateRangePreset { get; set; } = DateRangePreset.All;
    public DateTime CustomSince { get; set; } = DateTime.Today.AddMonths(-1);
    public DateTime CustomUntil { get; set; } = DateTime.Today;
    public HashSet<string> EnabledMessageTypes { get; set; } =
        MessageTypeFilterInfo.AllIds().ToHashSet(StringComparer.OrdinalIgnoreCase);
    public bool IncrementalExport { get; set; }
    public bool MapGroupNicknames { get; set; } = true;
    public bool EnableSpeechToText { get; set; }
    public HashSet<string> FavoriteIds { get; set; } = new(StringComparer.OrdinalIgnoreCase);

    public (DateTime? since, DateTime? until) ResolvedDateRange()
    {
        var now = DateTime.Now;
        return DateRangePreset switch
        {
            DateRangePreset.Last7Days => (now.AddDays(-7), now),
            DateRangePreset.Last30Days => (now.AddDays(-30), now),
            DateRangePreset.Last90Days => (now.AddDays(-90), now),
            DateRangePreset.Last365Days => (now.AddDays(-365), now),
            DateRangePreset.Custom => (CustomSince, CustomUntil),
            _ => (null, null)
        };
    }

    private static string SettingsPath =>
        System.IO.Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "WeChatExporter",
            "settings.json");

    private static string DefaultExportPath() =>
        System.IO.Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            "Downloads",
            "微信聊天记录导出");

    public void Save()
    {
        var dir = System.IO.Path.GetDirectoryName(SettingsPath)!;
        System.IO.Directory.CreateDirectory(dir);
        var json = JsonSerializer.Serialize(this, new JsonSerializerOptions { WriteIndented = true });
        System.IO.File.WriteAllText(SettingsPath, json);
    }

    private static AppSettings Load()
    {
        try
        {
            if (System.IO.File.Exists(SettingsPath))
            {
                var json = System.IO.File.ReadAllText(SettingsPath);
                var loaded = JsonSerializer.Deserialize<AppSettings>(json);
                if (loaded is not null)
                {
                    if (string.IsNullOrWhiteSpace(loaded.ExportPath))
                        loaded.ExportPath = DefaultExportPath();
                    if (loaded.CustomSince == default)
                        loaded.CustomSince = DateTime.Today.AddMonths(-1);
                    if (loaded.CustomUntil == default)
                        loaded.CustomUntil = DateTime.Today;
                    loaded.EnabledMessageTypes = NormalizeTypeSet(loaded.EnabledMessageTypes);
                    loaded.FavoriteIds = new HashSet<string>(
                        loaded.FavoriteIds.Where(id => !string.IsNullOrWhiteSpace(id)),
                        StringComparer.OrdinalIgnoreCase);
                    return loaded;
                }
            }
        }
        catch
        {
            // fall through to defaults
        }
        return new AppSettings();
    }

    private static HashSet<string> NormalizeTypeSet(HashSet<string>? raw)
    {
        var all = MessageTypeFilterInfo.AllIds().ToHashSet(StringComparer.OrdinalIgnoreCase);
        if (raw is null || raw.Count == 0)
            return all;

        var normalized = raw
            .Where(id => all.Contains(id))
            .ToHashSet(StringComparer.OrdinalIgnoreCase);
        return normalized.Count == 0 ? all : normalized;
    }
}

File: Sources/WeChatExporter/Models/AppSettings.swift
import Foundation
import SwiftUI

enum AppearanceMode: String, CaseIterable, Identifiable {
    case system
    case light
    case dark

    var id: String { rawValue }

    var title: String {
        switch self {
        case .system: return "跟随系统"
        case .light: return "浅色"
        case .dark: return "深色"
        }
    }

    var colorScheme: ColorScheme? {
        switch self {
        case .system: return nil
        case .light: return .light
        case .dark: return .dark
        }
    }
}

/// 应用设置（持久化到 UserDefaults）。
@MainActor
final class AppSettings: ObservableObject {
    static let shared = AppSettings()
    static let creditLine = "@林琝淏科技集团有限公司出品"

    @Published var appearance: AppearanceMode {
        didSet { defaults.set(appearance.rawValue, forKey: Keys.appearance) }
    }

    @Published var exportStyle: ExportStyle {
        didSet { defaults.set(exportStyle.rawValue, forKey: Keys.exportStyle) }
    }

    @Published var includeMedia: Bool {
        didSet { defaults.set(includeMedia, forKey: Keys.includeMedia) }
    }

    @Published var includeStickerGallery: Bool {
        didSet { defaults.set(includeStickerGallery, forKey: Keys.includeStickerGallery) }
    }

    @Published var folderIncludeCSV: Bool {
        didSet { defaults.set(folderIncludeCSV, forKey: Keys.folderIncludeCSV) }
    }

    @Published var folderIncludeJSON: Bool {
        didSet { defaults.set(folderIncludeJSON, forKey: Keys.folderIncludeJSON) }
    }

    @Published var openFolderAfterExport: Bool {
        didSet { defaults.set(openFolderAfterExport, forKey: Keys.openFolderAfterExport) }
    }

    @Published var exportPath: String {
        didSet { defaults.set(exportPath, forKey: Keys.exportPath) }
    }

    @Published var dateRangePreset: DateRangePreset {
        didSet { defaults.set(dateRangePreset.rawValue, forKey: Keys.dateRangePreset) }
    }

    @Published var customSince: Date {
        didSet { defaults.set(customSince.timeIntervalSince1970, forKey: Keys.customSince) }
    }

    @Published var customUntil: Date {
        didSet { defaults.set(customUntil.timeIntervalSince1970, forKey: Keys.customUntil) }
    }

    /// 启用的消息类型；空或全选 = 全部。
    @Published var enabledMessageTypes: Set<MessageTypeFilter> {
        didSet {
            defaults.set(enabledMessageTypes.map(\.rawValue).sorted(), forKey: Keys.enabledMessageTypes)
        }
    }

    @Published var incrementalExport: Bool {
        didSet { defaults.set(incrementalExport, forKey: Keys.incrementalExport) }
    }

    @Published var mapGroupNicknames: Bool {
        didSet { defaults.set(mapGroupNicknames, forKey: Keys.mapGroupNicknames) }
    }

    @Published var enableSpeechToText: Bool {
        didSet { defaults.set(enableSpeechToText, forKey: Keys.enableSpeechToText) }
    }

    @Published var favoriteIDs: Set<String> {
        didSet {
            defaults.set(Array(favoriteIDs).sorted(), forKey: Keys.favoriteIDs)
        }
    }

    private let defaults = UserDefaults.standard

    private enum Keys {
        static let appearance = "settings.appearance"
        static let exportStyle = "settings.exportStyle"
        static let includeMedia = "settings.includeMedia"
        static let includeStickerGallery = "settings.includeStickerGallery"
        static let folderIncludeCSV = "settings.folderIncludeCSV"
        static let folderIncludeJSON = "settings.folderIncludeJSON"
        static let openFolderAfterExport = "settings.openFolderAfterExport"
        static let exportPath = "settings.exportPath"
        static let dateRangePreset = "settings.dateRangePreset"
        static let customSince = "settings.customSince"
        static let customUntil = "settings.customUntil"
        static let enabledMessageTypes = "settings.enabledMessageTypes"
        static let incrementalExport = "settings.incrementalExport"
        static let mapGroupNicknames = "settings.mapGroupNicknames"
        static let enableSpeechToText = "settings.enableSpeechToText"
        static let favoriteIDs = "settings.favoriteIDs"
    }

    var resolvedDateRange: (since: Int?, until: Int?) {
        dateRangePreset.resolvedRange(customSince: customSince, customUntil: customUntil)
    }

    var filterOptions: ChatJsonProcessor.FilterOptions {
        ChatJsonProcessor.FilterOptions(
            sinceUnix: resolvedDateRange.since,
            untilUnix: resolvedDateRange.until,
            enabledTypes: enabledMessageTypes
        )
    }

    func isFavorite(_ id: String) -> Bool { favoriteIDs.contains(id) }

    func toggleFavorite(_ id: String) {
        if favoriteIDs.contains(id) { favoriteIDs.remove(id) }
        else { favoriteIDs.insert(id) }
    }

    func isTypeEnabled(_ type: MessageTypeFilter) -> Bool {
        enabledMessageTypes.isEmpty || enabledMessageTypes.contains(type)
    }

    func toggleType(_ type: MessageTypeFilter) {
        if enabledMessageTypes.isEmpty {
            enabledMessageTypes = Set(MessageTypeFilter.allCases)
        }
        if enabledMessageTypes.contains(type) {
            enabledMessageTypes.remove(type)
        } else {
            enabledMessageTypes.insert(type)
        }
        if enabledMessageTypes.count == MessageTypeFilter.allCases.count {
            // keep as full set for clarity
        }
    }

    private init() {
        let defaultExport = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Downloads/微信聊天记录导出", isDirectory: true)
            .path

        if let raw = defaults.string(forKey: Keys.appearance),
           let mode = AppearanceMode(rawValue: raw) {
            appearance = mode
        } else {
            appearance = .system
        }

        if let raw = defaults.string(forKey: Keys.exportStyle),
           let style = ExportStyle(rawValue: raw) {
            exportStyle = style
        } else {
            exportStyle = .singleHTML
        }

        includeMedia = defaults.object(forKey: Keys.includeMedia) as? Bool ?? false
        includeStickerGallery = defaults.object(forKey: Keys.includeStickerGallery) as? Bool ?? true
        folderIncludeCSV = defaults.object(forKey: Keys.folderIncludeCSV) as? Bool ?? true
        folderIncludeJSON = defaults.object(forKey: Keys.folderIncludeJSON) as? Bool ?? false
        openFolderAfterExport = defaults.object(forKey: Keys.openFolderAfterExport) as? Bool ?? false
        exportPath = defaults.string(forKey: Keys.exportPath) ?? defaultExport

        if let raw = defaults.string(forKey: Keys.dateRangePreset),
           let preset = DateRangePreset(rawValue: raw) {
            dateRangePreset = preset
        } else {
            dateRangePreset = .all
        }

        if defaults.object(forKey: Keys.customSince) != nil {
            customSince = Date(timeIntervalSince1970: defaults.double(forKey: Keys.customSince))
        } else {
            customSince = Calendar.current.date(byAdding: .month, value: -1, to: Date()) ?? Date()
        }
        if defaults.object(forKey: Keys.customUntil) != nil {
            customUntil = Date(timeIntervalSince1970: defaults.double(forKey: Keys.customUntil))
        } else {
            customUntil = Date()
        }

        if let raw = defaults.array(forKey: Keys.enabledMessageTypes) as? [String] {
            let set = Set(raw.compactMap(MessageTypeFilter.init(rawValue:)))
            enabledMessageTypes = set.isEmpty ? Set(MessageTypeFilter.allCases) : set
        } else {
            enabledMessageTypes = Set(MessageTypeFilter.allCases)
        }

        incrementalExport = defaults.object(forKey: Keys.incrementalExport) as? Bool ?? false
        mapGroupNicknames = defaults.object(forKey: Keys.mapGroupNicknames) as? Bool ?? true
        enableSpeechToText = defaults.object(forKey: Keys.enableSpeechToText) as? Bool ?? false
        if let ids = defaults.array(forKey: Keys.favoriteIDs) as? [String] {
            favoriteIDs = Set(ids)
        } else {
            favoriteIDs = []
        }
    }
}
